### PR TITLE
Allow for configuring a specific CML version

### PR DESCRIPTION
**Pull request by dacbd**
I haven't played with the task branch yet so, with those changes in mind how relevant would the following be?

I want to be able to target a specific cml version here: https://github.com/iterative/terraform-provider-iterative/blob/8121c664c4cc557059e44a332636d61a715c21dd/iterative/resource_runner.go#L288 similar to [this](https://github.com/iterative/setup-cml/pull/51)

----
adding an optional `cml_version` / defaulting to `@dvcorg/cml`
https://github.com/iterative/terraform-provider-iterative/blob/8121c664c4cc557059e44a332636d61a715c21dd/iterative/resource_runner.go#L409-L433

Thoughts @DavidGOrtega @0x2b3bfa0 ?

### CODE REVIEW

**Comment by casperdcl** [comment]
I feel like this sort of thing (deps & versions of deps) should be part of the user's startup script or task instead...

**Comment by 0x2b3bfa0** [comment]
This suggestion is not applicable to the `iterative_task` resource, because it isn't aware of CML in any way, and this would be handled on the users' startup scripts as @casperdcl points out.

Although, the legacy `iterative_cml_runner` **is** specific for CML runners and, as such, it would make sense to allow pinning a given version. 🤷🏼‍♂️

**Comment by 0x2b3bfa0** [comment]
Still, we[^1] chose to keep CML under the same minor version and avoid introducing breaking changes, so _in theory,_ there shouldn't be any need of pinning it.

[^1]: actually _they,_ in the year 1 before @0x2b3bfa0

**Comment by dacbd** [comment]
I'll play with task today, and I have a few ideas how I want to use it.

**Comment by DavidGOrtega** [comment]
@dacbd I think that might be very interesting. CML should be backwards compatible, at least thats the philosophy but since we allow to pin in `setup-cml` it should make sense here.

Just only one possible issue:
I thinks is a good idea replace CML npm package and node deps with the standalone CLI after #842

Any thoughts @dacbd @iterative/cml ?

**Comment by dacbd** [comment]
I think it depends on how it's executed. I haven't really experienced any issues with the node deps with CML but I am well aware of the issues the ecosystem can have, plus it makes installing/portability simpler which I feel is your guys targeted goal.

I see/have no issues against a standalone CLI. I personally feel that tools like these should have strong control/influence on their dependencies. I did a lot of work with the hapi.js framework a long time ago and their control over their dependencies offered a stable product than others.

You never know what bonehead doesn't have 2FA on their npm account and now your dependency has a crypto-miner or a package.json build script dialing home, granted these things get caught pretty fast these days, and realistically they normally have a low impact but the damage to the "name" is done anyway.

**Comment by 0x2b3bfa0** [comment]
https://github.com/iterative/cml/releases/latest/download/cml-linux

**Comment by dacbd** [comment]
I'll take a look, plan to incorporate something this weekend that you can take a look at.

**Comment by 0x2b3bfa0** [comment]
I'd even remove the “classical” NPM support and include only the required logic for downloading binary releases. Apart from that, looks good to me! 🎉 @iterative/cml?

**Comment by dacbd** [comment]
> I'd even remove the “classical” NPM support ...

This part of my covert 🔐 effort to make testing custom changes(in cml/tpi) easier (npm provides a convenient method for "installing" directly from a git ref), but that is secondary to actual function.

**Review by 0x2b3bfa0 — APPROVED** [review]

**Comment by 0x2b3bfa0** [comment]
Tip: `cd iterative && go test -update` to fix snapshot tests

**Comment by dacbd** [comment]
I will make a couple house cleaning changes tonight and mark it ready. (saw at least one empty line that was superfluous)

**Comment by dacbd** [comment]
Boy, can I say my brain wasn't working tonight for some reason...

After brute-forcing myself back to functional with some hacky e2e testing, I formed a couple questions:

Should this be updated further?
https://github.com/iterative/terraform-provider-iterative/blob/35ce161a602f7b6ad1b59ccf5c555bda5c66285e/iterative/resource_runner_test.go#L35-L70

Should Installing CML be part of `.setup` https://github.com/iterative/terraform-provider-iterative/blob/master/environment/setup.sh ?

Should that script actually log into that file? `/var/log/cml_stack.log`

----
I'm _fairly_ confident that I have unbroken all the things I broke.
hacky e2e test:
```
git clone git@github.com:dacbd/terraform-provider-iterative.git
pushd terraform-provider-iterative
git checkout target-cml-version
make build
make install
popd
git clone git@github.com:dacbd/cml.git
pushd cml
npm install
sed -i 's/source = "iterative\/iterative"/source = "github.com\/iterative\/iterative"/g' src/terraform.js
sed -i 's/return template;/console.log(template);return template;/g' src/terraform.js
bin/cml.js runner \
    --single \
    --token=***\
    --cloud=aws \
    --cloud-region=us-west-2 \
    --cloud-type=t3.micro \
    --cloud-startup-script=ZWNobyAiZWNkc2Etc2hhMi1uaXN0cDM4NCBBQUFBRTJWalpITmhMWE5vWVRJdGJtbHpkSEF6T0RRQUFBQUlibWx6ZEhBek9EUUFBQUJoQkRZZDNzc2E2TDE1alFDNWJja0oydmlXbEExdEJ5Z3hlV295M3MwUzE0WkhNeFVNZnA3dTJ5cWZpY3BITzViK3BqZ2c3THorODBJYncxNTd3YVRaUE0reGJGMi9LR3FTN2FZVjBML1I4VmJXalZFcHp4WkVlb3hTQ3dGQTF0SFdVZz09IGNhcmRubzowMDA2MTE4OTU1NzUiID4+IC9ob21lL3VidW50dS8uc3NoL2F1dGhvcml6ZWRfa2V5cwo=
popd
rm -rf cml
rm -rf terraform-provider-iterative
```

I tested it with aws/gcp (gcp lets you view the startup script the instance was made with which was nice)
I did **not** test all of the paths basically just `cml_version = ""` ie default/latest since there isn't a method in cml runner to set it.

<details>
<summary>lol wut google error?</summary>

I saw the instance appear with a spinning status in the GCP dashboard but they spit this error out...

https://gist.github.com/dacbd/41b9821c200dd38709a3f671b830ac49

</details>

**Comment by dacbd** [comment]
Made a `cml-version` [branch](https://github.com/iterative/cml/compare/master...dacbd:cml-version) to test additional values
- `cml_version = ""`
- `cml_version = "0.11.1"`
- `cml_version = "v0.11.1"`
- `cml_version = "git://github.com/dacbd/cml.git#cml-version" `
- `cml_version = "0.10.0"`
- `cml_version = "0.9.0"`
- `cml_version = "v0.8.1"`

**Review by 0x2b3bfa0 — APPROVED** [review]

**Comment by 0x2b3bfa0** [comment]
(Outdated confirmation)

> _Should this be updated further?_

Yes.

> _Should Installing CML be part of `.setup`?_

No, the setup section only takes care of Node.js, Docker and Nvidia drivers.

> _Should that script actually log into that file? `/var/log/cml_stack.log`_

No, despite the name, that file is just being used as a flag. 🙈

File: iterative/utils/helpers.go [hunk 1]
```diff
@@ -1,12 +1,81 @@
...
+		ver, err := semver.Make(version[1:])
+		if err == nil {
+			return getSemverCML(ver)
+		}
+	}
+	// handle semver
+	ver, err := semver.Make(version)
+	if err == nil {
+		return getSemverCML(ver)
+	}
```
Comment (0x2b3bfa0) [inline iterative/utils/helpers.go:41]:
```suggestion
	// handle semver
	ver, err := semver.Make(strings.TrimPrefix(ver, "v"))
	if err == nil {
		return getSemverCML(ver)
	}
```

File: iterative/utils/helpers.go [hunk 2]
```diff
@@ -1,12 +1,74 @@
...
+			return getNPMCML("@dvcorg/cml")
+		}
+		for _, asset := range release.Assets {
+			if *asset.Name == "cml-linux" {
+				return getGHCML(*asset.BrowserDownloadURL)
+			}
+		}
+	}
+	// handle semver
+	ver, err := semver.Make(strings.TrimPrefix(ver, "v"))
```
Comment (0x2b3bfa0) [inline iterative/utils/helpers.go:31]:
```suggestion
	ver, err := semver.Make(strings.TrimPrefix(version, "v"))
```

**Pull request merged**